File: api-java/samples/server/trinsic/StaticFileMiddleware.java
package id.trinsic;
import io.javalin.Javalin;
import io.javalin.http.Context;
import io.javalin.http.Handler;
import java.io.InputStream;
import java.io.FileInputStream;
import java.io.File;
import java.nio.file.Paths;
import java.io.Console;
public class StaticFileMiddleware implements Handler {

    private final String staticDir;

    public StaticFileMiddleware(String staticDir) {
        this.staticDir = staticDir;
    }

    @Override
    public void handle(Context ctx) throws Exception {
        String requestedPath = ctx.path(); // e.g. "/redirect"
        String filePath = Paths.get(staticDir, requestedPath).toString();

        if(requestedPath.equals("/")) {
            filePath = Paths.get(staticDir, "index.html").toString();
        }

        File file = new File(filePath);
        if (file.exists() && file.isFile()) {
            ctx.result(new FileInputStream(file));
            ctx.contentType(getMimeType(file.getName()));
            ctx.status(200);
            return;
        } else {
            filePath = Paths.get(staticDir, requestedPath + ".html").toString();
            file = new File(filePath);
            if (file.exists() && file.isFile()) {
                ctx.result(new FileInputStream(file));
                ctx.contentType(getMimeType(file.getName()));
                ctx.status(200);
                return;
            }
        }
    }
    private String getMimeType(String fileName) {
        String lower = fileName.toLowerCase();
        if (lower.endsWith(".html")) return "text/html";
        if (lower.endsWith(".css")) return "text/css";
        if (lower.endsWith(".js")) return "application/javascript";
        if (lower.endsWith(".json")) return "application/json";
        if (lower.endsWith(".png")) return "image/png";
        if (lower.endsWith(".jpg") || lower.endsWith(".jpeg")) return "image/jpeg";
        if (lower.endsWith(".svg")) return "image/svg+xml";
        return "application/octet-stream";
    }
}

File: api-php/samples/server/src/static-file-middleware.php
<?php
declare(strict_types=1);

namespace App\Middleware;

use Psr\Http\Message\ResponseInterface;
use Psr\Http\Message\ServerRequestInterface;
use Psr\Http\Server\MiddlewareInterface;
use Psr\Http\Server\RequestHandlerInterface;
use Slim\Psr7\Response;

class StaticFileMiddleware implements MiddlewareInterface
{
    private string $staticDir;

    public function __construct(string $staticDir)
    {
        $this->staticDir = rtrim($staticDir, DIRECTORY_SEPARATOR);
    }

    public function process(ServerRequestInterface $request, RequestHandlerInterface $handler): ResponseInterface
    {
        $path = $request->getUri()->getPath();
        if ($path === '/') {
            $path = '/index.html';
        }

        $filePath = realpath($this->staticDir . $path);

        if ($filePath !== false && str_starts_with($filePath, realpath($this->staticDir)) && is_file($filePath)) {
            $response = new Response();
            $mimeType = $this->getMimeType($filePath);
            $response = $response
                ->withHeader('Content-Type', $mimeType)
                ->withStatus(200);

            $stream = fopen($filePath, 'rb');
            $response->getBody()->write(stream_get_contents($stream));
            fclose($stream);

            return $response;
        }
        $filePath = realpath($this->staticDir . $path . '.html');
        error_log('[StaticFileMiddleware] Path: ' . $filePath);
        if ($filePath !== false && str_starts_with($filePath, realpath($this->staticDir)) && is_file($filePath)) {
            $response = new Response();
            $mimeType = $this->getMimeType($filePath);
            $response = $response
                ->withHeader('Content-Type', $mimeType)
                ->withStatus(200);

            $stream = fopen($filePath, 'rb');
            $response->getBody()->write(stream_get_contents($stream));
            fclose($stream);

            return $response;
        }

        // Fallback to default Slim handler
        return $handler->handle($request);
    }

    private function getMimeType(string $filePath): string
    {
        $ext = strtolower(pathinfo($filePath, PATHINFO_EXTENSION));
        return match ($ext) {
            'html' => 'text/html',
            'css' => 'text/css',
            'js' => 'application/javascript',
            'json' => 'application/json',
            'png' => 'image/png',
            'jpg', 'jpeg' => 'image/jpeg',
            'svg' => 'image/svg+xml',
            default => 'application/octet-stream',
        };
    }
}